Selection dialog redesign

itisAliRH opened the pull request:
This PR proposes a redesign of the `SelectionDialog`, aiming to enhance the user experience, match it with other parts, and improve code maintainability. Key changes include:

- **UI Improvements**: Updated styling and layout to provide a more intuitive and responsive user interface with a larger modal, including better handling of selection states and visual feedback. For example, prevent closing the modal by pressing the ESC button or clicking outside of the modal and losing selected items.
- **Enhanced Functionality**: Restrict the item's type and display relevant details using the type to improve usability, it added support for custom sorting keys and selection states (all, intermediate, none).
- **Refactoring and Cleanup**: Simplified the codebase by removing redundant fields and streamlining the selection dialog components. Needed more refactoring to the components using SelectionDialog in follow-up PRs to reduce redundant and outdated logic.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

itisAliRH committed:
Add sorting and list view types to `common/index` for better reusability

Expanded common components type definitions to include `ListView` and sorting options,

itisAliRH committed:
Remove unnecessary FontAwesome library imports and import missing bootstrap component in `ListHeader`

itisAliRH committed:
Add optional select-all checkbox to ListHeader

Enhanced the list header with a select-all checkbox to improve user experience when managing multiple items.

itisAliRH committed:
Add SelectionCard component for improved item selection

Introduced a new SelectionCard component to enhance user interaction with selection items. This component allows for better visual representation and management of items, including selection states and metadata display.

itisAliRH committed:
Remove unused leaf icon from DatasetCollectionDialog

itisAliRH committed:
Fixing SelectionDialog modal styling caused by overrides styles

Refactored the modal's SCSS incorporating breakpoints. This issue caused by forcing all modals to have limited width and max-width in overrides.scss

davelopez commented on client/src/components/SelectionDialog/SelectionCard.vue:
```diff
@@ -0,0 +1,224 @@
+                </div>
+
+                <div class="selection-card-footer">
+                    <StatelessTags
+                        v-if="(props.item.tags ?? []).length > 0"
+                        :value="props.item.tags"
+                        :disabled="true" />
```
```suggestion
                    <StatelessTags v-if="props.item.tags" :value="props.item.tags" :disabled="true" />
```

davelopez commented on client/src/components/SelectionDialog/selectionTypes.ts:
```diff
@@ -6,21 +6,19 @@ export const SELECTION_STATES = {
     isLeaf: boolean;
     url: string;
+    size?: number;
+    variant?: SelectionState;
+    tags?: string[];
     _rowVariant?: SelectionState;
```
What is the difference between `variant` and `_rowVariant`? can we use `selectionState?: SelectionState`?

davelopez commented on client/src/components/SelectionDialog/selectionTypes.ts:
```diff
@@ -6,21 +6,19 @@ export const SELECTION_STATES = {
     label: string;
-    details: string;
+    time?: Date | string;
+    update_time: Date | string;
```
I've seen this can be an empty string in some cases, can we instead make it optional and change it to `updateTime?: Date` so those empty strings are undefined. In general, we should have an "adapter" or mapping function that will translate the raw item to a `SelectionItem` so we can do the conversion/unification there. The same for `time` that can be maybe also called `creationTime` for consistency.

davelopez commented on client/src/components/SelectionDialog/SelectionDialog.vue:
```diff
@@ -44,21 +44,21 @@ interface Props {
+    sortKeys: () => defaultSortKeys,
     isBusy: false,
     isEncoded: false,
     items: () => [],
     itemsProvider: undefined,
```
I propose to drop the `items` property and manage the items always through the `itemsProvider` function this will eliminate some case handling and will allow proper pagination for all dialogs. It will also ensure proper mapping between the raw items and `SelectionItem`s.

ElectronicBlueberry replied:
I feel like the items provider mechanism could bite us down the road. It's a bit untransparent, and it seems like it could easily lead to bugs like overfetching. I'd rather we try to solve this with standard vue patterns like props+events or a store.

davelopez replied:
You are right, I was still thinking about the old `BTable`. In any case, we just need a common way of "providing" items to the dialog no matter where the data comes from (history, remote file sources, data libraries, etc.), with pagination, sorting and search support (where possible) and to normalize the data to the common `SelectionItem` interface,

davelopez reviewed (commented):
Looking pretty cool!

guerler commented on client/src/components/Common/ListHeader.vue:
```diff
@@ -1,25 +1,31 @@
-type SortBy = "create_time" | "update_time" | "name";
-
 interface Props {
+    sortKeys?: SortKey[];
```
Please sort props, computed values and function names alphabetically within their categories. It makes navigating the code easier.

guerler commented on client/src/components/Common/ListHeader.vue:
```diff
@@ -110,6 +114,8 @@ defineExpose({
 .list-header {
     display: flex;
+    width: 100%;
+    padding: 0 0.5rem 0.5rem;
```
Please avoid adding custom styles, we already have bootstrap, use d-flex, w-100, px or py.

guerler commented on client/src/components/SelectionDialog/SelectionCard.vue:
```diff
@@ -0,0 +1,224 @@
+    </div>
+</template>
+
+<style scoped lang="scss">
```
This is yet another card with custom styles and colors. There should be no custom color in the client code.

guerler commented on client/src/components/FilesDialog/FilesDialog.vue:
```diff
@@ -236,6 +224,8 @@ function open(record: SelectionItem) {
 
 /** Performs server request to retrieve data records **/
 function load(record?: SelectionItem) {
+    selectionDialog.value.resetPagination();
```
This needs to be called without using the ref to access member functions.

itisAliRH committed:
Refactor ListHeader component to only show sort options when available

itisAliRH committed:
Refactor selection item structure for improved flexibility

Updated the SelectionItem interface to enhance its adaptability by adding optional properties like `time`, `update_time`, `size`, `variant`, and `tags`. This change allows for more versatile data handling and better alignment with evolving requirements. Removed the `FieldEntry` interface to streamline the codebase.

itisAliRH committed:
Refactor SelectionDialog to use SelectionCard

Enhanced the SelectionDialog component by using new SelectionCard, icon imports, updating props for better sorting capabilities, and optimizing item fetching logic. These changes improve performance and user experience, particularly in handling large datasets. The UI has been refined for clarity, and unnecessary code has been removed to simplify maintenance.

itisAliRH committed:
Refactor FilesDialog for improved selection handling

Streamlined the selection dialog by removing unnecessary computed fields and resetting pagination on load.

itisAliRH committed:
Add mandatory update_time field to file source items in FIlesDialog

itisAliRH committed:
Refactor HistoryDatasetPicker

Defaulted the mandatory `update_time` property to an empty string.

itisAliRH committed:
Refactor BasicSelectionDialog types

Updated the `leafIcon` prop to use `IconDefinition`. Added a dedicated `onClick` function to handle item selection.

itisAliRH committed:
Enhance DataDialog with type safety and UI improvements

Refactored the DataDialog component to improve type safety by using `SelectionItem` for items and removed unnecessary fields definition. Imported  `FontAwesomeIcon` for icon rendering.

itisAliRH committed:
Improve pagination in SelectionDialog

Updated the watch function to include `currentPage` and `perPage` as dependencies, ensuring that filtered items are recalculated correctly when pagination changes when itemsProvider is undefined.

itisAliRH committed:
Refactor ListHeader to support dynamic sorting options

Updated the ListHeader component to utilize dynamic sorting keys. This change allows for easier adjustments to sorting criteria without modifying the core logic, addressing the need for a more adaptable UI.
